src/QsCompiler/QirGeneration/QIR/DataStructures.cs

```diff
@@ -386,85 +444,253 @@ private Value AllocateTuple(bool registerWithScopeManager)
     /// </summary>
     internal class ArrayValue : IValue
     {
+        // FIXME: ENFORCE THAT STACKALLOC IS ONLY TRUE WHEN ALL ELEMENTS ARE STACK ALLOC?
+        // TODO: make Paulis i2s rather than loading them.
         private readonly GenerationContext sharedState;
         private readonly IValue.Cached<Value> length;
+        private readonly IValue.Cached<PointerValue>[]? arrayElementPointers;
 
         internal ResolvedType QSharpElementType { get; }
 
-        public ITypeRef LlvmElementType { get; }
+        public ResolvedType QSharpType =>
+            ResolvedType.New(QsResolvedTypeKind.NewArrayType(this.QSharpElementType));
 
-        public uint? Count { get; }
+        public ITypeRef LlvmElementType { get; private set; }
 
-        public Value OpaquePointer { get; }
+        public ITypeRef LlvmType { get; private set; }
 
-        public Value Value => this.OpaquePointer;
+        public uint? Count { get; }
 
-        public ITypeRef LlvmType => this.sharedState.Types.Array;
+        public Value Value { get; private set; }
 
-        public ResolvedType QSharpType =>
-            ResolvedType.New(QsResolvedTypeKind.NewArrayType(this.QSharpElementType));
+        public Value OpaquePointer =>
+            Types.IsArray(this.LlvmType)
+            ? this.Value
+            : throw new InvalidOperationException("cannot get opaque pointer for a constant array allocated on the stack");
 
         public Value Length => this.length.Load();
 
         /// <summary>
         /// Creates a new array value.
+        /// The array will be stack allocated if and only if maxCount is set.
```

Comment:
It looks like this comment might be out of date, since the params are now `count` and `allocOnStack`

Reply:
Removed that line in particular, but I think I meant to make a general pass to update the doc comment to reflect my changes in the QIR generation. I'll go do that.